### PR TITLE
Mesh reader support for VTK high order elements [feature/vtk_high_order]

### PR DESCRIPTION
**Pull request by jonesholger**
# Updated PR information (Dec. 2020 – Will)

This PR adds two new features to MFEM's VTK support:

1. Reading XML VTK mesh files in addition to the "legacy" ASCII format
2. Reading high-order VTK meshes. High-order support works for **both** XML and legacy VTK formats.

XML parsing is done using the TinyXML2 library.

To facilitate systematic conversion between VTK's high-order node ordering and MFEM's node ordering, mappings are created that use an intermediate lexicographic ordering. Some remarks about the lexicographic ordering:

* All NodalFiniteElement subclasses now support GetLexicographicOrdering, which plays the same role as TensorBasisElement::GetDofMap, but with support for all non-tensor geometries
* The "refined geometry vertices" for tetrahedral elements are now lexicographically ordered in the standard reference element rather than in an "auxiliary element"
* As a consequence, the lexicographic DOF ordering given by GetLexicographicOrdering always agrees with the refined geometry vertex ordering.
* These orderings are tested for all geometries, for orders 1 through 6 in the unit tests.
* The VTK mesh reader supports arbitrary-order Lagrange elements. The node orderings are handled using the same mappings defined for high-order ParaView output, so pre-generated data tables are not required.

As part of this PR, the handling of the VTK format for input and output are both made more systematic. The files `vtk.hpp` and `vtk.cpp` now have a lot more features that should make dealing with VTK more convenient, for instance converting from a VTK element type to MFEM `Geometry::Type` can now be performed using `VTKGeometry:: GetMFEMGeometry` instead of manual switch statements. This simplifies the code in `Mesh::PrintVTK` and `Mesh::PrintVTU`.

For testing purposes, VTK meshes of Cartesian grids with low-order and high-order elements of all types can be created in ParaView with the source "Unstructured Cell Types".

# Previous information (out of date)

> This mesh reader update subtask for vtk is related to MFEM issue https://github.com/mfem/mfem/issues/678  & https://blog.kitware.com/modeling-arbitrary-order-lagrange-finite-elements-in-the-visualization-toolkit/
> 
> Quads to arbitrary order vtk ascii
> Hexes to arbitrary order vtk ascii
> Triangles to order 6 (vtk limit) vtk ascii
> Tets to order 6 (vtk limit) vtk ascii
> Wedge (or prism) to order 6 (vtk limit) vtk ascii
> Create static maps based on Reference Elements up to order 6 for tri/tet/wedge
> Mixed cell validation -- assuming all elements are the newer Lagrange High Order Elements
> 
> Samples have been uploaded to https://github.com/mfem/data/tree/master/vtk; more will be added later

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#975](https://github.com/mfem/mfem/pull/975) | @jonesholger | @tzanio | @acfisher + @bslazarov + @mlstowell | 11/30/20 | 01/09/21 | 01/24/21 | |

### CODE REVIEW

**Comment by jonesholger** [comment]
Some background links
Consolidating a few related links here:
https://github.com/GLVis/glvis/issues/64
https://blog.kitware.com/modeling-arbitrary-order-lagrange-finite-elements-in-the-visualization-toolkit/
 https://glvis.org/curvilinear-vtk-meshes/
https://github.com/CEED/FMS
https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf
https://lorensen.github.io/VTKExamples/site/VTKFileFormats/
https://www.visitusers.org/index.php?title=ASCII_VTK_Files
https://github.com/lorensen/VTKExamples
https://vtk.org/doc/nightly/html/vtkCellType_8h_source.html
https://github.com/agrayver/VTK_Lagrange
https://github.com/Kitware/VTK/blob/265ca48a79a36538c95622c237da11133608bbe5/Common/DataModel/vtkLagrangeQuadrilateral.cxx#L558
https://github.com/Kitware/VTK/blob/265ca48a79a36538c95622c237da11133608bbe5/Common/DataModel/vtkLagrangeHexahedron.cxx#L734
https://github.com/Kitware/VTK/blob/675adbc0feeb3f62730ecacb2af87917af124543/Filters/Sources/vtkCellTypeSource.cxx#L1307
https://github.com/Kitware/VTK/blob/675adbc0feeb3f62730ecacb2af87917af124543/Filters/Sources/vtkCellTypeSource.cxx#L345
https://github.com/Kitware/VTK/blob/5efd279913b4842bed0c1a0f2f119db89134e378/Common/DataModel/vtkCellType.h#L107-L114
https://github.com/Kitware/VTK/blob/675adbc0feeb3f62730ecacb2af87917af124543/Filters/Sources/vtkCellTypeSource.cxx#L1463
https://www.paraview.org/Wiki/ParaView_Release_Notes#Arbitrary-order_Lagrange_finite_element_support
https://github.com/Kitware/VTK/blob/675adbc0feeb3f62730ecacb2af87917af124543/Common/DataModel/vtkLagrangeInterpolation.cxx#L208
https://github.com/mfem/mfem/blob/3e9a86cecfd422027da0f44c671cf2c630206074/fem/fe.cpp#L6878-L6917
https://gitlab.kitware.com/vtk/vtk/blob/master/Common/DataModel/vtkUnstructuredGrid.h#L339
 https://github.com/dealii/dealii/commit/b029f83d2c2b2877b26503ab060717354bd02668
https://github.com/dealii/dealii/commit/e65f90698bf5c14964bf76534f0ac48e0632819f
 https://kitware.github.io/paraview-docs/latest/python/paraview.simple.UnstructuredCellTypes.html

File: mesh/mesh_readers.cpp [hunk 1]
```diff
@@ -670,7 +1035,7 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
                }
             }
          }
-      }
+         } // for NumOfElements
```
Comment (acfisher) [inline mesh/mesh_readers.cpp]: Is the whitespace/tabing off here?

File: mesh/mesh_readers.cpp [hunk 2]
```diff
@@ -685,10 +1050,184 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
             }
          }
       }
+         points.Destroy();
```
Comment (acfisher) [inline mesh/mesh_readers.cpp]: Maybe some more whitespaec weirdness.  You should be able to run make style on the branch to fix stuff like this up.

**Comment by pazner** [comment]
We just discussed this feature this week. We should revive this PR.

**Comment by pazner** [comment]
I just pushed some updates to this branch:

* All `NodalFiniteElement` subclasses now support `GetLexicographicOrdering`, which plays the same role as `TensorBasisElement::GetDofMap`, but with support for all non-tensor geometries
* The "refined geometry vertices" for tetrahedral elements are now lexicographically ordered in the standard reference element rather than in an "auxiliary element"
* As a consequence, the lexicographic DOF ordering given by `GetLexicographicOrdering` always agrees with the refined geometry vertex ordering. This could be useful for LOR on non-tensor-product elements. This is also always the same as the L2 DOF ordering.
* These orderings are tested for all geometries, for orders 1 through 6 in the unit tests.
* The VTK mesh reader supports arbitrary-order Lagrange elements. The node orderings are handled using the same mappings defined for high-order ParaView output, so pre-generated data tables are not required.

**Comment by tzanio** [comment]
☝️ very nice work @pazner! Is this close to be ready for review?

**Comment by pazner** [comment]
> ☝️ very nice work @pazner! Is this close to be ready for review?

I think this is very close. I'm waiting to see if the latest commit fixed AppVeyor. It would also be good to test the mesh reader on more complicated high-order VTK meshes, but they're a little hard to come by. It works on all the example meshes in the VTK subdirectory of the data repo (except for LagrangeHexWedge.vtk, which I think is malformed), but those are all fairly trivial.

**Comment by pazner** [comment]
CI looks good now. See also #1894 for reading XML VTK files.

Feedback and testing would be helpful :)

**Comment by pazner** [comment]
To help the reviewers, I edited the [top-level comment](https://github.com/mfem/mfem/pull/975#issue-291267443) with an updated summary of this PR — the information there was previously quite out of date.

**Comment by tzanio** [comment]
@acfisher, @bslazarov and @mlstowell -- can you please review?

File: tests/unit/fem/test_lexicographic_ordering.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+template <typename T> void VerifyOrdering(int order)
+{
+   T el(order, BasisType::GaussLobatto);
+   Geometry::Type geom = el.GetGeomType();
+   INFO("order " << order << " " << Geometry::Name[geom]);
+   VerifyOrdering(el);
+}
+
+TEST_CASE("Lexicographic Ordering", "[FiniteElement,Geometry]")
```
Comment (acfisher) [inline tests/unit/fem/test_lexicographic_ordering.cpp:48]: This is a good test to have around!

File: fem/fe.cpp [hunk 4]
```diff
@@ -8525,61 +8539,85 @@ H1_TetrahedronElement::H1_TetrahedronElement(const int p, const int btype)
    Vector shape_x(p + 1), shape_y(p + 1), shape_z(p + 1), shape_l(p + 1);
 #endif
 
+   auto tri = [](int k) { return (k*(k + 1))/2; };
```
Comment (acfisher) [inline fem/fe.cpp:8542]: Good use for lambdas.

**Review by acfisher — APPROVED** [review]

File: fem/fe.hpp [hunk 5]
```diff
@@ -731,6 +732,15 @@ class NodalFiniteElement : public ScalarFiniteElement
    virtual void ProjectDiv(const FiniteElement &fe,
                            ElementTransformation &Trans,
                            DenseMatrix &div) const;
+
+   /** @brief Get an Array<int> that maps lexicographically ordered indices to
```
Comment (mlstowell) [inline fem/fe.hpp:736]: I think we should define exactly what is meant by "lexicographically ordered".  Without knowing the lexicon and the rules we use to order it this array may not be useful.

**Comment by tzanio** [comment]
Extending approval due date to the end of this week...

File: fem/fe.hpp (hunk 5 above)
Comment (pazner) [reply fem/fe.hpp:736]: How does the updated comment look to you?

Comment (mlstowell) [reply fem/fe.hpp:736]: Looks perfect. Thanks!

**Review by mlstowell — APPROVED** [review]

**Comment by tzanio** [comment]
ping: @bslazarov

**Review by bslazarov — APPROVED** [review]

**Comment by tzanio** [comment]
Can you please update `CHANGELOG`

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Review by tzanio — APPROVED** [review]
@jonesholger and @pazner -- thank you so much for this important contributions. I am sure many of our users will appreciate it!

@mlstowell, @acfisher and @bslazarov -- thank you for reviewing!

**Pull request merged**